url587829ef81_champdavis_KhovanovHomology_tgz-pJ8-KhovanovMapJ8/src/MergeMap.java

```diff
@@ -0,0 +1,77 @@
+import java.util.*;
+
+public class MergeMap implements DMap{
+       
+    private KhovanovMapNode fromNode;
+    private KhovanovMapNode toNode;
+    private ArrayList<SeifertCircle> fromCircles;
+    private SeifertCircle toCircle;
+
+    public MergeMap(KhovanovMapNode f, KhovanovMapNode t, ArrayList<SeifertCircle> fc, SeifertCircle tc){
+        this.fromNode = f;
+        this.toNode = t;
+        this.fromCircles = fc;
+        this.toCircle = tc;  
+        
+        addTensorArrows();
+        System.out.println(toString());
+    }
+    
+    public KhovanovMapNode getToNode(){
+        return this.toNode;
+    }
+    
+    public void addTensorArrows(){
+        Tensor[] from = this.fromNode.getBasis();
+        Tensor[] to = this.toNode.getBasis();
+        
+        /*   V+V+ |--> V+
+         *   V+V- |--> V-
+         *   V-V+ |--> V-
+         *   V-V- |--> 0
+         */  
+        BasisElement toVminus = this.toCircle.getVminus();
+        BasisElement toVplus = this.toCircle.getVplus();
+        
+        BasisElement fromCircOneVminus = this.fromCircles.get(0).getVminus();
+        BasisElement fromCircOneVplus = this.fromCircles.get(0).getVplus();
+        BasisElement fromCircTwoVminus = this.fromCircles.get(1).getVminus();
+        BasisElement fromCircTwoVplus = this.fromCircles.get(1).getVplus();
+        
+        Arrow theArrow;
+        
+        for(Tensor f:from){
+            if(f.contains(fromCircOneVplus) && f.contains(fromCircTwoVplus)){          
+                for(Tensor t:to){
+                    
+                    if(f.isIdentityMinusOne(t,toVplus)){
+
+                        if(t.contains(toVplus)){
+                            theArrow = new Arrow(f,t);
+                            f.addOutArrow(theArrow);
+                            t.addInArrow(theArrow);  
+                        }
+                    }
+                }        
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 45:55, 58:68.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.